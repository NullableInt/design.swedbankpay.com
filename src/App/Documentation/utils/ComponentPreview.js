import React from "react";
import PropTypes from "prop-types";
import { renderToStaticMarkup } from "react-dom/server";
import PrismCode from "react-prism";
import jsbeautifier from "js-beautify";

const ComponentPreview = ({ children, language, removeOuterTag, showCasePanel, codeFigure }) => {
    // TODO: This is stupid, find a better way to do this [EH]
    // should be possible with React 16.2
    // https://stackoverflow.com/questions/33766085/how-to-avoid-extra-wrapping-div-in-react
    const _removeOuterTag = element => {
        const div = document.createElement("div");
        div.innerHTML = renderToStaticMarkup(element);

        if (div.firstElementChild) {
            return div.firstElementChild.innerHTML;
        } else if (div.firstChild) {
            return div.firstChild.nodeValue;
        }
        return "Check ComponentPreview _removeOuterTag!";
    };

    const CodeFigure = () => {
        let code = "";

        if (language === "html" && children && typeof children.map === "function") {
            children.map(child => {
                if (removeOuterTag) {
                    code += _removeOuterTag(child);
                } else {
                    code += renderToStaticMarkup(child);
                }
            });

        } else if (language === "html") {
            if (removeOuterTag) {
                code += _removeOuterTag(children);
            } else {
                code += renderToStaticMarkup(children);
            }
        } else {
            switch (typeof children) {
                case "string":
                    code = children;
                    break;
                case "object":
                    children.forEach(child => code += child);
                    break;
                default:
                    console.warn("CodeFigure: children needs attention!");
            }
        }

        switch (language) {
            case "html":
                code = jsbeautifier.html_beautify(code);
                code = code.replace(/=""/g, "");
                break;
            case "css":
                code = jsbeautifier.css_beautify(code);
                break;
            case "javascript":
                code = jsbeautifier.js_beautify(code, {
                    brace_style: "preserve-inline", // eslint-disable-line camelcase
                    e4x: true
                });
                break;
            default:
                return "update switchcase!";
        }

        return (
            <figure>
                <pre>
                    <PrismCode className={`language-${language}`}>
                        {code}
                    </PrismCode>
                </pre>
            </figure>
        );
    };


    const ShowCasePanel = () => (
        <div className="showcase-panel">
            {children}
        </div>
    );

    return (
        <div>
            {showCasePanel ? <ShowCasePanel /> : null}
            {codeFigure ? <CodeFigure /> : null}
        </div>
    );
};

ComponentPreview.propTypes = {
    language: PropTypes.string.isRequired,
    removeOuterTag: PropTypes.bool,
    showCasePanel: PropTypes.bool,
    codeFigure: PropTypes.bool
};

const Attribute = ({ data, name, value }) => (
    <code>
        {data ? <span className="token attr-name">data-px-</span> : null}
        <span className="token attr-name">{name}</span>
        <span className="token attr-value">
            <span className="token punctuation">{"=\""}</span>
            {value}
            <span className="token punctuation">{"\""}</span>
        </span>
    </code>
);

Attribute.propTypes = {
    name: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    data: PropTypes.bool
};

const PxScript = ({ component, subComponents, func, params }) => {
    let renderedParams, renderedSubComponents;

    if (subComponents) {
        renderedSubComponents = subComponents.map((component, i) => (
            <span key={i}>{component}
                <span className="token punctuation">.</span>
            </span>
        ));
    }

    if (params) {
        renderedParams = params.map((param, i) => (
            <span key={i}>{param}
                {(i < params.length - 1) ? <span className="token punctuation">, </span> : null}
            </span>
        ));
    }

    return (
        <code>
            <span>px</span>
            <span className="token punctuation">.</span>
            <span>{component}</span>
            <span className="token punctuation">.</span>
            {renderedSubComponents}
            <span className="token function">{func}</span>
            <span className="token punctuation">(</span>
            {renderedParams}
            <span className="token punctuation">);</span>
        </code>
    );
};

PxScript.propTypes = {
    component: PropTypes.string.isRequired,
    subComponents: PropTypes.array,
    func: PropTypes.string.isRequired,
    params: PropTypes.array
};

export default ComponentPreview;

export { Attribute, PxScript };
