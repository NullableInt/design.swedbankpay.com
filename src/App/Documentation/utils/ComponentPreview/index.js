import React from "react";
import PropTypes from "prop-types";
import { renderToStaticMarkup } from "react-dom/server";
import PrismCode from "react-prism";
import jsbeautifier from "js-beautify";

// NOTE: dangerousHTML prop is used when wanting to show html in the codefigure without encoding.

const ComponentPreview = ({ children, language, removeOuterTag, hideValue, removeList, showCasePanel, codeFigure, dangerousHTML }) => {
    const _removeOuterTag = element => {
        const div = document.createElement("div");
        div.innerHTML = renderToStaticMarkup(element);

        if (div.firstElementChild) {
            return div.firstElementChild.innerHTML;
        } else if (div.firstChild) {
            return div.firstChild.nodeValue;
        }
        return "Check ComponentPreview _removeOuterTag!";
    };

    const _removeList = element => {
        const div = document.createElement("div");
        div.innerHTML = renderToStaticMarkup(element);

        const listElements = div.querySelectorAll("li");
        let htmlString = "";

        listElements.forEach(el => {
            htmlString += `${el.innerHTML} \n`;
        });

        return htmlString;
    };

    const _removeEmpty = val => {
        val = val.replace(/=""/g, "");
        val = val.replace(/ class>/g, ">");
        val = val.replace(/ class /g, " ");
        val = val.replace(/value /g, "");

        return val;
    };

    const setDangerousHtml = val => {
        let code = "";
        if (typeof val.forEach === "function") {
            val.forEach(v => code += `${v}\n`);
        } else {
            code = val;
        }
        return code;
    };

    const CodeFigure = () => {
        let code = "";

        if (language === "html" && dangerousHTML) {
            code = setDangerousHtml(children);
        } else if (language === "html" && children && typeof children.map === "function") {
            children.map(child => {
                if (removeOuterTag) {
                    code += _removeOuterTag(child);
                } else if (removeList) {
                    code += _removeList(child);
                } else {
                    code += renderToStaticMarkup(child);
                }
            });

        } else if (language === "html") {
            if (removeOuterTag) {
                code += _removeOuterTag(children);
            } else if (removeList) {
                code += _removeList(children);
            } else {
                code += renderToStaticMarkup(children);
            }
        } else {
            switch (typeof children) {
                case "string":
                    code = children;
                    break;
                case "object":
                    children.forEach(child => code += child);
                    break;
                default:
                    console.warn("CodeFigure: children needs attention!");
            }
        }

        switch (language) {
            case "html":
                code = jsbeautifier.html_beautify(code);
                code = _removeEmpty(code);
                if (hideValue) {
                    code = code.replace(/ value="(.*)"/g, "");
                }
                break;
            case "css":
                code = jsbeautifier.css_beautify(code);
                break;
            case "javascript":
                code = jsbeautifier.js_beautify(code, {
                    brace_style: "preserve-inline", // eslint-disable-line camelcase
                    e4x: true
                });
                break;
            default:
                return "update switchcase!";
        }

        return (
            <figure>
                <pre>
                    <PrismCode className={`language-${language}`}>
                        {code}
                    </PrismCode>
                </pre>
            </figure>
        );
    };


    const ShowCasePanel = () => (
        <div className="showcase-panel">
            {children}
        </div>
    );

    return (
        <>
            {showCasePanel ? <ShowCasePanel /> : null}
            {codeFigure ? <CodeFigure /> : null}
        </>
    );
};

ComponentPreview.propTypes = {
    language: PropTypes.oneOf(["html", "javascript", "css"]).isRequired,
    removeOuterTag: PropTypes.bool,
    hideValue: PropTypes.bool,
    removeList: PropTypes.bool,
    showCasePanel: PropTypes.bool,
    codeFigure: PropTypes.bool,
    dangerousHTML: PropTypes.bool
};

export default ComponentPreview;
