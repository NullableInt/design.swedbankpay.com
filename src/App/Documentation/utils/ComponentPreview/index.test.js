import React from "react";
import { mount, shallow } from "enzyme";
import "prismjs";

import ComponentPreview from "./index";

// const htmlEncodeString = str => String(str)
//     .replace(/&/g, "&amp;")
//     .replace(/</g, "&lt;")
//     .replace(/>/g, "&gt;")
//     .replace(/"/g, "&quot;");

describe("Utilities: ComponentPreview", () => {
    const TestComponentH1 = () => (<h1>test1</h1>);
    const TestComponentH2 = () => (<h2>test2</h2>);
    const TestComponentH1WithOuterTags = () => (
        <div>
            <h1>test1</h1>
        </div>
    );

    it("is defined", () => {
        expect(ComponentPreview).toBeDefined();
    });

    // it("renders", () => {
    //     const wrapper = mount(
    //         <ComponentPreview language="html" showCasePanel={true} codeFigure={true}>
    //             <TestComponentH1 />
    //         </ComponentPreview>
    //     );

    //     expect(wrapper).toMatchSnapshot();
    //     expect(wrapper.contains(<h1>test1</h1>)).toEqual(true);
    //     expect(wrapper.html()).toContain("<figure>");
    //     expect(wrapper.html()).toContain("showcase-panel");
    // });

    it("does not render codeFigure when prop is false/not provided", () => {
        const wrapper = mount(
            <ComponentPreview language="html" showCasePanel={true}>
                <TestComponentH1 />
            </ComponentPreview>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.html()).not.toContain("<figure>");
    });

    it("does not render showCasePanel when prop is false/not provided", () => {
        const wrapper = mount(
            <ComponentPreview language="html" codeFigure={true}>
                <TestComponentH1 />
            </ComponentPreview>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.html()).not.toContain("showcase-panel");
    });

    it("CodeFigure renders multiple html tags", () => {
        const wrapper = mount(
            <ComponentPreview language="html" codeFigure={true}>
                <TestComponentH1 />
                <TestComponentH2 />
            </ComponentPreview>
        );

        expect(wrapper).toMatchSnapshot();
        // TODO: Update expects [EH]

        // expect(wrapper.html()).toContain(htmlEncodeString("<h1>test1</h1>"));
        // expect(wrapper.html()).toContain(htmlEncodeString("<h2>test2</h2>"));
        // expect(wrapper.html("<h2>test2</h2>")).toEqual(true);
    });

    it("CodeFigure removes outer tags from component", () => {
        const wrapper = mount(
            <ComponentPreview language="html" codeFigure={true} removeOuterTag={true}>
                <TestComponentH1WithOuterTags />
                <TestComponentH1WithOuterTags />
            </ComponentPreview>
        );

        expect(wrapper).toMatchSnapshot();
        // TODO: Update expects [EH]

        // expect(wrapper.contains(<div><h1>test1</h1></div>)).toEqual(true);
    });

    it("CodeFigure warns about unhandled children", () => {
        console.warn = jest.fn();

        const wrapper = mount(
            <ComponentPreview language="css" codeFigure={true}>
                {() => []}
            </ComponentPreview>
        );

        expect(wrapper).toMatchSnapshot();
        expect(console.warn).toHaveBeenCalled();
    });

    it("CodeFigure renders css string", () => {
        const wrapper = mount(
            <ComponentPreview language="css" codeFigure={true}>
                {"color: red;"}
            </ComponentPreview>
        );

        // TODO: Update expects [EH]
        expect(wrapper).toMatchSnapshot();
    });

    it("CodeFigure renders multiple javascript strings", () => {
        const wrapper = mount(
            <ComponentPreview language="javascript" codeFigure={true}>
                {"const a = 12;"}
                {"const b = 1;"}
                {"const c = a + b;"}
            </ComponentPreview>
        );

        // TODO: Update expects [EH]
        expect(wrapper).toMatchSnapshot();
    });

    // it("prop language other than javascript, css or html returns warning message", () => {
    //     const wrapper = mount(
    //         <ComponentPreview language="test" codeFigure={true}>
    //             {"test"}
    //         </ComponentPreview>
    //     );

    //     // TODO: Update expects [EH]
    //     expect(wrapper).toMatchSnapshot();
    //     expect(wrapper.html()).toContain("update switchcase!");
    // });
});

