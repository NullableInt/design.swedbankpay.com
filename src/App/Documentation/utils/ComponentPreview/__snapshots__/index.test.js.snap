// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utilities: ComponentPreview CodeFigure removes outer tags from component 1`] = `
<ComponentPreview
  codeFigure={true}
  language="html"
  removeOuterTag={true}
>
  <CodeFigure>
    <figure>
      <pre>
        <PrismCode
          className="language-html"
          component="code"
        >
          <code
            className="language-html"
          >
            &lt;h1&gt;test1&lt;/h1&gt;
&lt;h1&gt;test1&lt;/h1&gt;
          </code>
        </PrismCode>
      </pre>
    </figure>
  </CodeFigure>
</ComponentPreview>
`;

exports[`Utilities: ComponentPreview CodeFigure renders css string 1`] = `
<ComponentPreview
  codeFigure={true}
  language="css"
>
  <CodeFigure>
    <figure>
      <pre>
        <PrismCode
          className="language-css"
          component="code"
        >
          <code
            className="language-css"
          >
            color: red;
          </code>
        </PrismCode>
      </pre>
    </figure>
  </CodeFigure>
</ComponentPreview>
`;

exports[`Utilities: ComponentPreview CodeFigure renders multiple html tags 1`] = `
<ComponentPreview
  codeFigure={true}
  language="html"
>
  <CodeFigure>
    <figure>
      <pre>
        <PrismCode
          className="language-html"
          component="code"
        >
          <code
            className="language-html"
          >
            &lt;h1&gt;test1&lt;/h1&gt;
&lt;h2&gt;test2&lt;/h2&gt;
          </code>
        </PrismCode>
      </pre>
    </figure>
  </CodeFigure>
</ComponentPreview>
`;

exports[`Utilities: ComponentPreview CodeFigure renders multiple javascript strings 1`] = `
<ComponentPreview
  codeFigure={true}
  language="javascript"
>
  <CodeFigure>
    <figure>
      <pre>
        <PrismCode
          className="language-javascript"
          component="code"
        >
          <code
            className="language-javascript"
          >
            const a = 12;const b = 1;const c = a + b;
          </code>
        </PrismCode>
      </pre>
    </figure>
  </CodeFigure>
</ComponentPreview>
`;

exports[`Utilities: ComponentPreview CodeFigure warns about unhandled children 1`] = `
<ComponentPreview
  codeFigure={true}
  language="css"
>
  <CodeFigure>
    <figure>
      <pre>
        <PrismCode
          className="language-css"
          component="code"
        >
          <code
            className="language-css"
          />
        </PrismCode>
      </pre>
    </figure>
  </CodeFigure>
</ComponentPreview>
`;

exports[`Utilities: ComponentPreview does not render codeFigure when prop is false/not provided 1`] = `
<ComponentPreview
  language="html"
  showCasePanel={true}
>
  <ShowCasePanel>
    <div
      className="showcase-panel"
    >
      <TestComponentH1>
        <h1>
          test1
        </h1>
      </TestComponentH1>
    </div>
  </ShowCasePanel>
</ComponentPreview>
`;

exports[`Utilities: ComponentPreview does not render showCasePanel when prop is false/not provided 1`] = `
<ComponentPreview
  codeFigure={true}
  language="html"
>
  <CodeFigure>
    <figure>
      <pre>
        <PrismCode
          className="language-html"
          component="code"
        >
          <code
            className="language-html"
          >
            &lt;h1&gt;test1&lt;/h1&gt;
          </code>
        </PrismCode>
      </pre>
    </figure>
  </CodeFigure>
</ComponentPreview>
`;
